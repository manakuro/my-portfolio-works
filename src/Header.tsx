import * as React from 'react';
import * as Swiper from 'react-id-swiper';

export default class Header extends React.Component {
  private swiper: any;

  constructor(props: {}) {
    super(props);
    this.goNext = this.goNext.bind(this);
    this.goPrev = this.goPrev.bind(this);
  }

  public render(): JSX.Element {
    const params = {
      freeMode: true,
      loop: true,
      slidesPerView: 'auto',
      spaceBetween: 10,
    };

    const languages: string[] = [
      'Ruby on Rails',
      'React.js',
      'Angular',
      'Vue.js',
      'Fuel PHP',
      'TypeScript',
      'GraphQL',
      'JEST',
      'PWAs',
      'SSR',
      'Elm',
    ];

    const rows: JSX.Element[] = languages.map((l, index) => (
      <a href="#" key={index}>
        {l}
      </a>
    ));

    return (
      <header className="header">
        <div className="logo">
          <a href="#">Manato</a>
        </div>
        <nav className="nav">
          <div>
            <i className="fa fa-angle-left icon" onClick={this.goPrev} />
          </div>
          <div className="nav-container">
            <Swiper
              {...params}
              ref={(node: any) => {
                if (node) {
                  this.swiper = node.swiper;
                }
              }}
            >
              {rows}
            </Swiper>
          </div>
          <div>
            <i className="fa fa-angle-right icon" onClick={this.goNext} />
          </div>
        </nav>
      </header>
    );
  }

  private goNext(): void {
    this.swiper.slideNext();
  }

  private goPrev(): void {
    this.swiper.slidePrev();
  }
}
