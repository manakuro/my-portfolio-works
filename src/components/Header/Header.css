.header {
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100px;
  background: rgba(255, 255, 255, 0.9);
  display: flex;
  flex-direction: column;
  justify-content: center; }
  .header .logo {
    display: flex;
    align-items: center;
    justify-content: center;
    height: 40px; }
    .header .logo > a {
      font-size: 2rem;
      font-weight: 700;
      font-family: 'Source Serif Pro', serif;
      color: #2F2F2F; }
  .header .nav {
    display: flex;
    justify-content: center;
    align-items: center; }
    .header .nav /deep/ .swiper-slide {
      width: auto; }
    .header .nav .nav-container {
      width: 50%;
      overflow: hidden; }
      .header .nav .nav-container a {
        font-family: 'Roboto';
        font-size: .9rem;
        color: #2F2F2F;
        opacity: .7;
        border-bottom: 1px solid transparent;
        transition: .25s; }
        .header .nav .nav-container a.active {
          opacity: 1;
          color: #06dca3;
          border-bottom: 1px solid #05f9b8b5; }
        .header .nav .nav-container a:hover {
          opacity: 1; }
    .header .nav .icon {
      cursor: pointer;
      padding: 0 10px; }
