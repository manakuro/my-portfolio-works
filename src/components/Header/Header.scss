$headerHeight: 100px;

.header {
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: $headerHeight;
  background: rgba(255, 255, 255, 0.9);
  display: flex;
  flex-direction: column;
  justify-content: center;

  .logo {
    display: flex;
    align-items: center;
    justify-content: center;
    height: 40px;

    > a {
      font-size: 2rem;
      font-weight: 700;
      font-family: 'Source Serif Pro', serif;
      color: #2F2F2F;
    }
  }

  .nav {
    display: flex;
    justify-content: center;
    align-items: center;

    /deep/ .swiper-slide {
      width: auto;
    }

    .nav-container {
      width: 50%;
      overflow: hidden;

      a {
        font-family: 'Roboto';
        font-size: .9rem;
        font-weight: 600;
        color: #2F2F2F;
        opacity: .7;
        border-bottom: 1px solid transparent;

        &.active {
          opacity: 1;
          color: #06dca3;
          border-bottom: 1px solid #05f9b8b5;
        }

        transition: .25s;

        &:hover {
          opacity: 1;
        }
      }
    }


    // angle icon
    .icon {
      cursor: pointer;
      padding: 0 10px;
    }
  }
}


