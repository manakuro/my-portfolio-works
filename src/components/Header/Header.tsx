import * as React from 'react'
import * as Swiper from 'react-id-swiper'
import * as queryString from 'query-string'
import { xor } from 'lodash'
import * as classNames from 'classnames'
import compose from 'recompose/compose'

import './Header.css'
import { HomeState, SearchQuery } from '@/modules/home/reducer'
import { History } from 'history'
import { lifecycle, onlyUpdateForKeys, withHandlers } from 'recompose'

export interface HeaderProps {
  languages: HomeState['languages']
  searchQuery: SearchQuery
  updateSearchQuery: (searchQuery: any) => any
  history: History
}

export interface WithHandlers {
  handleClick: (e: React.SyntheticEvent<EventTarget>, id: number) => any
}

const SWIPER_OPTIONS = {
  freeMode: true,
  slidesPerView: 'auto',
  spaceBetween: 10,
}

export function search(props: HeaderProps): void {
  props.history.push({
    pathname: '/',
    search: `?${queryString.stringify(props.searchQuery)}`,
  })
}

export function handleClick(
  e: React.SyntheticEvent<EventTarget>,
  props: HeaderProps,
  id: number,
): void {
  e.preventDefault()

  const languages = xor(props.searchQuery.languages, [id])
  props.updateSearchQuery({ languages })
}

export function next(swiper: any): void {
  swiper.slideNext()
}

export function prev(swiper: any): void {
  swiper.slidePrev()
}

const enhance = compose<HeaderProps & WithHandlers, HeaderProps>(
  withHandlers({
    handleClick: (props: HeaderProps) => async (
      e: React.SyntheticEvent<EventTarget>,
      id: number,
    ) => handleClick(e, props, id),
  }),
  lifecycle({
    componentWillReceiveProps(nextProps: HeaderProps) {
      if (this.props.searchQuery.languages !== nextProps.searchQuery.languages)
        search(nextProps)
    },
  }),
  onlyUpdateForKeys([
    'languages',
    'searchQuery',
    'updateSearchQuery',
    'history',
  ]),
)

export const Header = (props: HeaderProps & WithHandlers): JSX.Element => {
  let swiper: any

  const mapActiveLanguages = props.searchQuery.languages.reduce((acc, l) => {
    acc[l] = true
    return acc
  }, {})

  const rows: JSX.Element[] | null = props.languages.length
    ? props.languages.map(l => (
        // @todo figure out the best practice for passing args in event handler
        // @see https://github.com/palantir/tslint-react/issues/96
        <a
          href="#"
          key={l.id}
          // tslint:disable-next-line jsx-no-lambda
          onClick={e => props.handleClick(e, l.id)}
          className={classNames('header-Nav_Link', {
            'header-Nav_Link-active': mapActiveLanguages[l.id],
          })}
        >
          {l.name}
        </a>
      ))
    : null

  const nav: JSX.Element | null = rows ? (
    <Swiper
      {...SWIPER_OPTIONS}
      ref={(node: any) => {
        if (node) swiper = node.swiper
      }}
    >
      {rows}
    </Swiper>
  ) : null

  return (
    <header className="header">
      <div className="header-Logo">
        <a href="/">Manato</a>
      </div>
      <nav className="header-Nav">
        <div>
          <i
            className="fa fa-angle-left header-Nav_Icon"
            // tslint:disable-next-line jsx-no-lambda
            onClick={() => prev(swiper)}
          />
        </div>
        <div className="header-Nav_Container">{nav}</div>
        <div>
          <i
            className="fa fa-angle-right header-Nav_Icon"
            // tslint:disable-next-line jsx-no-lambda
            onClick={() => next(swiper)}
          />
        </div>
      </nav>
    </header>
  )
}

export default enhance(Header)
