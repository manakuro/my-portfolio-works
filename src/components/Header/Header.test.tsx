import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Header } from './Header'
import { HeaderProps, WithHandlers } from '@/components/Header/Header'

const props: HeaderProps & WithHandlers = {
  languages: [],
  history: {
    push: jest.fn(),
  } as any,
  updateSearchQuery: jest.fn(),
  searchQuery: {
    languages: [],
  },
  handleClick: jest.fn(),
}

describe('Header', () => {
  it('should render', () => {
    const wrapper = shallow(<Header {...props} />)

    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
