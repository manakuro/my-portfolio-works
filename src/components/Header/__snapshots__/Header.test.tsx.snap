// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render 1`] = `
<header
  className="header"
>
  <div
    className="header-Logo"
  >
    <a
      href="/"
    >
      Manato
    </a>
  </div>
  <nav
    className="header-Nav"
  >
    <div>
      <i
        className="fa fa-angle-left header-Nav_Icon"
        onClick={[Function]}
      />
    </div>
    <div
      className="header-Nav_Container"
    />
    <div>
      <i
        className="fa fa-angle-right header-Nav_Icon"
        onClick={[Function]}
      />
    </div>
  </nav>
</header>
`;
